adaptdl/adaptdl/goodput.py

```diff
@@ -0,0 +1,256 @@
+
+    def efficiency(self, batch_size):
+        grad_sqr = self._grad_params['norm']
+        grad_var = self._grad_params['var']
+        scale = batch_size * self._init_batch_size
+        denom = grad_var / scale + grad_sqr
+        gain = np.where(denom > 0, (grad_var + grad_sqr) / denom, 1.0)
+        return gain / scale
+
+    def optimize(self, num_nodes, num_replicas, max_batch_size=None,
+                 atomic_bsz_range=None, accumulation=False):
+        assert np.all(np.less_equal(1, num_nodes))
+        assert np.all(np.less_equal(num_nodes, num_replicas))
+        if max_batch_size is None:
+            max_batch_size = self._init_batch_size
+        assert self._init_batch_size <= max_batch_size
+        atomic_bsz_range = atomic_bsz_range or (None, None)
+        min_atomic_bsz = atomic_bsz_range[0] or 1
+        max_atomic_bsz = atomic_bsz_range[1] or max_batch_size
+        # Remember what the output shape/format should be and flatten inputs.
+        output_shape = np.broadcast(num_nodes, num_replicas).shape
+        output_scalar = np.isscalar(num_nodes) or np.isscalar(num_replicas)
+        num_nodes = np.broadcast_to(num_nodes, output_shape).flatten()
+        num_replicas = np.broadcast_to(num_replicas, output_shape).flatten()
+        # Samples 50 different total batch sizes in geometric space.
+        batch_size = np.geomspace(self._init_batch_size, max_batch_size)
+        local_bsz = np.expand_dims(batch_size, -1) / num_replicas
+        if accumulation:
+            # If local_bsz size exceeds the max atomic batch size, split it
+            # into a number of batches to form (atomic_bsz, accum_steps) such
+            # that (atomic_bsz * (accum_steps + 1)) is close to local_bsz.
+            #
+            # If num_replicas == 1 and local_bsz > self._init_batch_size, then
+            # always set accum_steps to 1. This is because (1) the gradient
+            # statistics used for scaling up the learning rate are inaccurate
+            # when there is only one atomic minibatch to estimate them from,
+            # and (2) using a lower accum_steps should always yield a higher
+            # goodput when there is only a single replica.
+            accum_steps = np.ceil(local_bsz / max_atomic_bsz) - 1
+            accum_steps = np.where(
+                (num_replicas == 1) & (local_bsz > self._init_batch_size),
+                np.maximum(accum_steps, 1), accum_steps).astype(np.int)
+        else:
+            accum_steps = np.zeros_like(local_bsz, dtype=np.int)
+        atomic_bsz = np.ceil(local_bsz / (accum_steps + 1)).astype(np.int)
+        # Evaluate the goodput of all candidate configurations.
+        goodput = self.evaluate(num_nodes, num_replicas,
+                                atomic_bsz, accum_steps)
+        # Set the goodput of invalid configurations to 0.0.
+        goodput = np.where((min_atomic_bsz <= atomic_bsz) &
+                           (atomic_bsz <= max_atomic_bsz), goodput, 0.0)
```

Comment:
I'm not a huge fan of replacing all invalid batchsizes afterwards, because it could be the case that geomspace missed all the valid configurations.

Reply:
For example, if replicas is 1, the init_batch_size is 1000, the minimum atomic bsz is 900, the maximum atomic bsz is 1010, and the max_batch_size is 1000000, geomspace(1000, 1000000) will produce only invalid total batch sizes, despite 1900 being a valid total batch size.